Add Rule to check for singular `yield` in `@contextlib.{async}contextmanager` decorated functions (RUF062)

maxmynter opened the pull request:
<!--
Thank you for contributing to Ruff! To help us out with reviewing, please consider the following:

- Does this pull request include a summary of the change? (See below.)
- Does this pull request include a descriptive title?
- Does this pull request include references to any relevant issues?
-->
Closes #16949 

## Summary
Add a rule that checks that only a single `yield` statement is in `@contextlib.contextmanager` or `@contextlib.asynccontextmanager` decorated functions. 

Later `yield` statements are executed in the cleanup stage of the contextmanager and result in a `RuntimeError`.

See [docs](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager)

### Notes
- I reused the `function_def_visit_preorder_except_body` from the `unused_async.rs` rule and moved it to `helpers.rs` alongside `class_def_visit_preorder_except_body`. While I don't use the latter, I moved both to maintain code locality.



<!-- What's the purpose of the change? What does it do, and why? -->

## Test Plan
Added a testing fixture and `cargo insta` snapshot.
<!-- How was it tested? -->

maxmynter committed:
(RUF060) Only single yield in @contextlib.contextmanager fn body

maxmynter committed:
(refactor) Move {fn,class}_def_visit_preorder_except body to helpers

as the fn version is reused in the context manager multiple yields
RUF060.

The class traversal isn't yet, but it makes sense to maintain locality.

maxmynter committed:
Register RUF060

maxmynter committed:
(tests, RUF060) Only single yield in contextlib.contextmanager

maxmynter committed:
Update Schema

maxmynter commented:
I've looked through these ecosystem changes.
They all come from branches in the control flow like this one:

```python
@contextmanager
def context():
   print("setup")
   if condition:
       print("condition")
       yield
   else:
       print("not condition")
       yield
    print("teardown")
```

The current implementation does not check for this; it should. I'm working on it.

maxmynter committed:
(tests) Extend and organize RUF060 tests

maxmynter committed:
(WIP) RUF060 should traverse execution path

Currently fails at properly handling try-except-else clauses.

maxmynter commented:
Still work in progress. Currently handling of `try-except-else` blocks is wrong and I also want to extend the test for deeper nested cases.

maxmynter committed:
(test, fixup) Fix test case typo

maxmynter committed:
(WIP) Refactor Node traversal to use Stmt introspection

and override visit for Try. This may be important for match too.

Will check that tomorrow.

maxmynter committed:
(test) Reorganize and extend for RUF060

Move into two files to seperate valid and invalid cases.

maxmynter committed:
(wip) Fix else clauses for loops and try blocks

maxmynter committed:
Factor out similar post-processing logic

maxmynter committed:
(refactor) Encapsulate yield stack top increment

maxmynter committed:
Improve error handling in RUF060

maxmynter committed:
Cleanup

ntBre commented:
I think you've picked a really tough rule to implement because it requires such a careful analysis of control flow. I added a minimized version of the first ecosystem hit as a test case:

```python
@contextlib.contextmanager
def foo():
    if False:
        yield 1
        return
    yield 2
```

and it is a real false positive.

I think this rule's implementation might be simplified by the control flow graph work being tracked in #17065, so I'd probably recommend holding off on this for now.

Thanks for your work here and apologies for not noticing this or commenting on the issue earlier!

For now, the main other thing you could do is just go through the ecosystem results and see if there are any other interesting test cases. The two I randomly picked also had this early return form, so they may make up a majority of false positives.

maxmynter commented:
I went through the ecosystem changes. Most are the return case you describe. Sometimes a little more complex: 

```python
try:
    yield
finally:
    return
```
So an implementation that tracks returns over control flow branches would also need to consider cases like this one where the return does not follow in the same block.


There are [two](https://github.com/ibis-project/ibis/blob/21e47ac0b39dc07122b67a95f0e92abd162a746f/ibis/backends/athena/__init__.py#L286) [other](https://github.com/zulip/zulip/blob/440864e7c92511f1181899133d5cc87b4988dfdc/zerver/lib/context_managers.py#L36) try except blocks like this:

```python
try:
    yield 1
except:
    yield 2
```

which in principle is a true positive. If the `try` fails after the `yield`, there would be a second one in the `except` block.

maxmynter committed:
(tests) Add tests for false positives in ecosystem changes

maxmynter committed:
Respect early returns in control flow yield count

ntBre commented:
Thanks again for working on this! I just want to make sure you saw this part of my comment above since I saw some new commits:

> I think this rule's implementation might be simplified by the control flow graph work being tracked in https://github.com/astral-sh/ruff/issues/17065, so **I'd probably recommend holding off on this for now**.

You're certainly welcome to keep working on it, but I think it will be easier to implement and to review once we have more general control-flow graph support.

cc @dylwil3 to make sure I'm not totally off-base with what the CFG work will give us

maxmynter committed:
(test) Extend test cases

maxmynter committed:
(tests) Adapt tests to assume try failure before yield

maxmynter committed:
Adapt RUF060 to assume failure in `try` before `yield`

ntBre commented:
Thanks Dylan!

I'd still lean towards waiting on the separate CFG work, and then the rule will grow capabilities as the CFG infrastructure improves, but I'm open to other thoughts too. The extra unit tests from the ecosystem check will still be very helpful for knowing where we stand throughout the process.

maxmynter committed:
(tests) Add tests for except return handling

maxmynter commented:
Awesome! Looking forward! 

I'll mark this as draft until then and this probably doesn't need a review before the CFG is ready (unless of course someone is bored; in that case i'd appreciate any pointers to improve my Rust).

maxmynter commented:
Note: The merge conflicts are with #16480 which by now has been merged as `RUF060`.
A renaming here should do. I'm Happy to take care once we decide to move forward here 

I would love to get this one over the finish line sometime and given extensive test cases and the ecosystem checks, I'm also pretty confident. The actual implementation is only 350ish lines. Does it still make sense to wait on the CFG work, @ntBre?

ntBre commented:
Sure, I can try to review it at some point if you fix the conflicts.

ntBre commented:
Just as a note on the rule code, you may want to jump to RUF062. https://github.com/astral-sh/ruff/pull/17368 is currently using RUF061 too.

maxmynter commented:
Awesome, I will tidy up here over the next few days and reopen and request a review once I'm there. 

@dylwil3 I hope I didn't come off as pushy and if so, apologies. I guess, I'm just excited about this contribution and definitely understand that you all have other priorities as weil!

maxmynter committed:
Merge branch 'main' into ctxmg-yield-once

maxmynter committed:
(fixup) Update Rule name reference in comment

maxmynter committed:
(refactor) Naming & Style

maxmynter committed:
Remove unreferenced Snapshot

that was stale from merge

ntBre committed:
Merge branch 'main' into ctxmg-yield-once

ntBre committed:
fix deeper merge conflicts

ntBre reviewed (commented):
Thanks. I haven't really wrapped my head fully around the stack stuff, but I have a few suggestions that might make things more clear.

maxmynter commented on crates/ruff_linter/resources/test/fixtures/ruff/RUF062.py:
```diff
@@ -0,0 +1,350 @@
+# (21) Valid: Return in finally guards second yield
+@contextlib.contextmanager
+def valid_try_except_yield_finally_return():
+    def is_true():
+        # Need this to avoid unreachable last yield
+        return True
+    print("Setting up")
+    if is_true():
```
Agree. Back when I was writing these, I wasn't too deep in Ruff's testing setup and got intimidated when the linting of the fixture began screaming at me! Removed these

maxmynter committed:
Review Comments on tests

maxmynter commented:
Test todos:
- [ ] Review Comments on tests
- [ ] Rename fixture `RUF062.py` -> `RUF062_2.py` 
- [ ] Update Snapshots

Rule Logic Todos
- [ ] Add tests for break
- [ ] Implement handling for `break`
- [ ] Review comments on logic

maxmynter committed:
Remove unreferenced snapshot

maxmynter commented on crates/ruff_linter/src/rules/ruff/rules/helpers.rs:
```diff
@@ -209,3 +210,62 @@ pub(super) fn is_descriptor_class(func: &Expr, semantic: &SemanticModel) -> bool
     })
 }
+
+/// Very nearly `crate::node::StmtFunctionDef.visit_preorder`, except it is specialized and,
```
It was changed here 

```bash
commit 050f33277163db62449d11e9cc84137be26661d4 refs/remotes/upstream/dhruv/callable-disjoint
Author: Micha Reiser <micha@reiser.io>
Date:   Fri Mar 28 20:40:26 2025 +0100

    Rename `visit_preorder` to `visit_source_order` (#17046)

    ## Summary

    We renamed the `PreorderVisitor` to `SourceOrderVisitor` a long time ago
    but it seems that we missed to rename the `visit_preorder` functions to
    `visit_source_order`.
    This PR renames `visit_preorder` to `visit_source_order`
```
with this being the (generated) code: 
https://github.com/astral-sh/ruff/blob/e6fe2af292e7be87c2917863464a5455c5b22c38/crates/ruff_python_ast/src/generated.rs#L7288-L7321

ntBre replied:
Looks like everything is updated except the comment 😆 Thanks for finding the commit!

maxmynter committed:
Rename *_visit_preorder_* -> *_visit_source_order_*

maxmynter committed:
Discriminate between {async}contextmanager in error message

maxmynter commented on crates/ruff_linter/src/rules/ruff/rules/multiple_yields_contextmanager.rs:
```diff
@@ -0,0 +1,373 @@
+
+    if path_tracker.has_multiple_yields {
+        checker.report_diagnostic(MultipleYieldsInContextManager, function_def.identifier());
+    }
```
No. This is never transiently true, so this is a great idea!

maxmynter committed:
Use HashMap to track violating Yield Expressions instead of count

We need a HashMap to track because a yield may be on several control
flow paths and thus added repeatedly to the violations.

Using the yield expressions we can later report multiple diagnostics
for a single violation. Currently, we report on the offending yields but
with a diagnostic for every particular one.

maxmynter committed:
Clippedyclip

maxmynter committed:
(refactor) Merge returns and yields into Scope struct

(refactor) Clean up naming

maxmynter committed:
Use .expect instead of debug_assert!

maxmynter committed:
(refactor) Simplify code built around solved borrowing issues

maxmynter committed:
Add docs to YieldPathVisitor

maxmynter committed:
(fixup) Max except yield path detection

maxmynter committed:
(refactor) debug_assert => expect

maxmynter committed:
Use shared slice instead of Vec

ntBre reviewed (commented):
I think the tests look good, but I'm hoping we can still simplify the code a bit more.

maxmynter commented:
Thanks @ntBre for your review. I'm somewhat occupied the next two weeks but will continue here afterwards!

maxmynter committed:
Resolve merge conflicts (excl. snapshot updates)

This still contains failing tests of RUFF062.
Updates in how output is formatted requires new snapshots.
This will be done in a separate PR.

maxmynter committed:
Update Cargo Insta Snapshot

maxmynter committed:
Extend documentation

maxmynter commented on crates/ruff_linter/src/rules/ruff/rules/multiple_yields_contextmanager.rs:
```diff
@@ -0,0 +1,489 @@
+use crate::{FixAvailability, Violation};
+
+/// ## What it does
+/// Checks that a function decorated with `contextlib.contextmanager` yields at most once.
```
Adressed in [7781d14](https://github.com/astral-sh/ruff/pull/17281/commits/7781d140633b62716cd06cc5795e1287b74bd7c1)

maxmynter committed:
Use static string slice allocation

maxmynter committed:
Refactor violation to use collection pattern

maxmynter committed:
Refactor to remove repated violation checks

maxmynter commented on crates/ruff_linter/src/rules/ruff/rules/multiple_yields_contextmanager.rs:
```diff
@@ -0,0 +1,489 @@
+                MultipleYieldsInContextManager::new(context_manager_name.clone()),
+                expr.range(),
+            );
+        }
```
Thanks for your suggestion. I've incorporated parts of it in [395e8ad](https://github.com/astral-sh/ruff/pull/17281/commits/395e8ad8139113570a893447761eb31d042d0d14) but tried to keep the concerns of the visitor and caller separate.

maxmynter committed:
Refactor to avoid loop check duplication

maxmynter committed:
(chore) Reduce duplication and improve naming

maxmynter committed:
(chore) Shorten Variable names and apply helper consistently

maxmynter committed:
(chore) Make pop_scope Option for tailored expect

maxmynter committed:
(chore) Remove clone() operations

maxmynter committed:
(chore) Break down try stmt handling into smaller functions

maxmynter committed:
(chore) Cleanup, Comments, and Reordering

maxmynter committed:
(chore) Remove unused self by associating function

maxmynter committed:
Create Struct for Try-Except paths handling

maxmynter committed:
(fixup) Stale comment

maxmynter commented:
Finally got around to do some more work here! Thanks (again) for your review @ntBre. 

I've 
- Resolved merge conflicts. Just renaming of methods.
- Broken up the Try-except-else-finally logic into smaller functions to make it easier to reason about
- use a collection-emission pattern for the yield counting.
- Improved naming quite a bit
- Added a couple more brief comments
- Removed a couple unnecessary allocations
- Extracted a couple code duplications

Let me know what you think.

ntBre reviewed (commented):
Thanks for following up on this and continuing to handle my pedantic feedback :)

All of my minor concerns are resolved, but I'm still thinking the implementation is more complicated than it needs to be. This is kind of a generalization of my comment below, but I just started deleting chunks of code and all of our tests still pass. Here's the patch I was working on locally:

<details><summary>Patch</summary>

```diff
diff --git a/crates/ruff_linter/src/rules/ruff/rules/multiple_yields_contextmanager.rs b/crates/ruff_linter/src/rules/ruff/rules/multiple_yields_contextmanager.rs
index 305b21d3ed..fc73075430 100644
--- a/crates/ruff_linter/src/rules/ruff/rules/multiple_yields_contextmanager.rs
+++ b/crates/ruff_linter/src/rules/ruff/rules/multiple_yields_contextmanager.rs
@@ -66,18 +66,8 @@ impl Violation for MultipleYieldsInContextManager {
 /// RUF062
 pub(crate) fn multiple_yields_in_contextmanager(checker: &Checker, function_def: &StmtFunctionDef) {
     if let Some(context_manager_name) = get_contextmanager_decorator(function_def, checker) {
-        let mut violations = Vec::new();
-        {
-            let mut yield_tracker = YieldTracker::new(&mut violations);
-            source_order::walk_body(&mut yield_tracker, &function_def.body);
-        }
-
-        for range in violations {
-            checker.report_diagnostic(
-                MultipleYieldsInContextManager::new(context_manager_name),
-                range,
-            );
-        }
+        let mut yield_tracker = YieldTracker::new(checker, context_manager_name);
+        source_order::walk_body(&mut yield_tracker, &function_def.body);
     }
 }
 
@@ -107,16 +97,18 @@ fn get_contextmanager_decorator(
 // Within a scope we evaluate all control flow paths and propagate the yields along the
 // maximum path to the outer scope.
 // Return exits the contextmanager decorated function and we stop accumulating yields along that path.
-struct YieldTracker<'a> {
-    violations: &'a mut Vec<TextRange>,
+struct YieldTracker<'a, 'b> {
+    checker: &'a Checker<'b>,
+    name: &'static str,
     scopes: Vec<YieldScope<'a>>,
     reported_ranges: FxHashSet<TextRange>,
 }
 
-impl<'a> YieldTracker<'a> {
-    fn new(violations: &'a mut Vec<TextRange>) -> Self {
+impl<'a, 'b> YieldTracker<'a, 'b> {
+    fn new(checker: &'a Checker<'b>, name: &'static str) -> Self {
         Self {
-            violations,
+            checker,
+            name,
             scopes: vec![YieldScope::new()],
             reported_ranges: FxHashSet::default(),
         }
@@ -124,51 +116,36 @@ impl<'a> YieldTracker<'a> {
 
     fn add_yield(&mut self, expr: &'a Expr) {
         if let Some(scope) = self.scopes.last_mut() {
-            scope.add_yield(expr);
+            scope.yield_expressions.push(expr);
             if scope.yields_excessively() {
                 self.emit_violation(expr);
             }
         }
     }
 
-    fn emit_violation(&mut self, expr: &'a Expr) {
+    fn emit_violation(&mut self, expr: &Expr) {
         let range = expr.range();
         if self.reported_ranges.insert(range) {
-            self.violations.push(range);
+            self.checker
+                .report_diagnostic(MultipleYieldsInContextManager::new(self.name), range);
         }
     }
 
-    fn emit_multiple_violations(&mut self, yields: &[&'a Expr]) {
+    fn emit_multiple_violations<'e>(&mut self, yields: impl IntoIterator<Item = &'e &'e Expr>) {
         // The first yield conforms to the protocol
-        for &yield_expr in yields.iter().skip(1) {
+        for yield_expr in yields.into_iter().skip(1) {
             self.emit_violation(yield_expr);
         }
     }
 
-    fn report_excess(&mut self, yields: &[&'a Expr]) {
-        if yields.len() > 1 {
-            self.emit_multiple_violations(yields);
-        }
-    }
-
     fn propagate_yields(&mut self, yields: &[&'a Expr]) {
-        self.report_excess(yields);
-        let scope = self
-            .scopes
-            .last_mut()
-            .expect("Missing current scope for yield propagation");
         for &yield_expr in yields {
-            scope.add_yield(yield_expr);
-        }
-        let yields_excessive = scope.yields_excessively();
-        let yield_exprs_clone = scope.yield_expressions.clone();
-        if yields_excessive {
-            self.emit_multiple_violations(&yield_exprs_clone);
+            self.add_yield(yield_expr);
         }
     }
 
-    fn push_scope(&mut self, scope: YieldScope<'a>) {
-        self.scopes.push(scope);
+    fn push_scope(&mut self) {
+        self.scopes.push(YieldScope::new());
     }
 
     fn pop_scope(&mut self) -> Option<(Vec<&'a Expr>, bool)> {
@@ -177,18 +154,11 @@ impl<'a> YieldTracker<'a> {
             .map(|scope| (scope.yield_expressions, scope.does_return))
     }
 
-    fn clear_scope_yields(&mut self) {
-        self.scopes
-            .last_mut()
-            .expect("Missing current scope for clearing yields")
-            .clear();
-    }
-
-    fn max_yields(branches: &[Vec<&'a Expr>]) -> Vec<&'a Expr> {
+    fn max_yields<'v>(branches: &'v [Vec<&'a Expr>]) -> &'v [&'a Expr] {
         branches
             .iter()
             .max_by_key(|branch| branch.len())
-            .cloned()
+            .map(Vec::as_slice)
             .unwrap_or_default()
     }
 
@@ -200,7 +170,7 @@ impl<'a> YieldTracker<'a> {
 
     fn handle_loop(&mut self, body: &'a [ast::Stmt], orelse: &'a [ast::Stmt]) {
         self.visit_body(body);
-        self.push_scope(YieldScope::new());
+        self.push_scope();
         self.visit_body(orelse);
     }
 
@@ -220,7 +190,7 @@ impl<'a> YieldTracker<'a> {
             let (except_yields, except_returns) = self
                 .pop_scope()
                 .expect("Missing except handler scope in try-statement");
-            self.report_excess(&except_yields);
+            self.emit_multiple_violations(&except_yields);
 
             if except_returns {
                 returning_except_branches.push(except_yields);
@@ -233,10 +203,6 @@ impl<'a> YieldTracker<'a> {
             .pop_scope()
             .expect("Missing try block scope in try-statement");
 
-        self.report_excess(&try_yields);
-        self.report_excess(&else_yields);
-        self.report_excess(&finally_yields);
-
         let path = TryExceptPath {
             try_yields,
             try_returns,
@@ -248,27 +214,14 @@ impl<'a> YieldTracker<'a> {
         };
 
         if finally_returns {
-            self.handle_terminating_paths(&path);
         } else {
             self.handle_continuing_paths(&path);
         }
     }
 
-    // Finally returns - execution stops, report worst-case path
-    fn handle_terminating_paths(&mut self, path: &TryExceptPath<'a>) {
-        let except_yields = Self::get_max_except_path(path);
-        let base_path = Self::build_pre_finally_path(path, &except_yields);
-        let max_path = Self::append_finally(&base_path, &path.finally_yields);
-
-        self.report_excess(&max_path);
-        self.clear_scope_yields();
-    }
-
     // Finally doesn't return - execution continues, handle all paths
     fn handle_continuing_paths(&mut self, path: &TryExceptPath<'a>) {
-        let (exception_return, exception_no_return) = Self::build_except_paths(path);
-
-        self.report_excess(&exception_return);
+        let (_exception_return, exception_no_return) = Self::build_except_paths(path);
 
         let normal_path = Self::build_try_else_path(path);
 
@@ -276,60 +229,23 @@ impl<'a> YieldTracker<'a> {
     }
 
     fn handle_exclusive_branches(&mut self, branch_count: usize) {
-        let mut returning_branches = Vec::new();
         let mut continuing_branches = Vec::new();
 
         for _ in 0..branch_count {
             let (branch_yields, branch_returns) = self
                 .pop_scope()
                 .expect("Missing branch scope in if/match statement");
-            self.report_excess(&branch_yields);
 
-            if branch_returns {
-                returning_branches.push(branch_yields);
-            } else {
+            if !branch_returns {
                 continuing_branches.push(branch_yields);
             }
         }
 
-        let max_returning = Self::max_yields(&returning_branches);
         let max_continuing = Self::max_yields(&continuing_branches);
 
-        self.report_excess(&max_returning);
         self.propagate_yields(&max_continuing);
     }
 
-    // Path building methods
-    // Find except handler with most yields
-    fn get_max_except_path(path: &TryExceptPath<'a>) -> Vec<&'a Expr> {
-        let max_returning_except = Self::max_yields(&path.returning_except_branches);
-        let max_continuing_except = Self::max_yields(&path.continuing_except_branches);
-
-        if max_returning_except.len() > max_continuing_except.len() {
-            max_returning_except
-        } else {
-            max_continuing_except
-        }
-    }
-
-    // Build path before finally: try + (else or except)
-    fn build_pre_finally_path(
-        path: &TryExceptPath<'a>,
-        max_except_yields: &[&'a Expr],
-    ) -> Vec<&'a Expr> {
-        let mut common_path = path.try_yields.clone();
-
-        if !path.try_returns {
-            common_path.extend(if path.else_yields.len() > max_except_yields.len() {
-                &path.else_yields
-            } else {
-                max_except_yields
-            });
-        }
-
-        common_path
-    }
-
     // Build exception paths with finally appended
     fn build_except_paths(path: &TryExceptPath<'a>) -> (Vec<&'a Expr>, Vec<&'a Expr>) {
         let max_returning_except = Self::max_yields(&path.returning_except_branches);
@@ -357,17 +273,15 @@ impl<'a> YieldTracker<'a> {
         exception_no_return: &[&'a Expr],
     ) {
         if path.try_returns {
-            let try_path = Self::append_finally(&path.try_yields, &path.finally_yields);
-            self.report_excess(&try_path);
             self.propagate_yields(exception_no_return);
         } else if path.else_returns {
-            self.report_excess(normal_path);
+            self.emit_multiple_violations(normal_path);
             self.propagate_yields(exception_no_return);
         } else {
             let max_yield_path = if normal_path.len() > exception_no_return.len() {
-                normal_path.to_vec()
+                normal_path
             } else {
-                exception_no_return.to_vec()
+                exception_no_return
             };
             self.propagate_yields(&max_yield_path);
         }
@@ -397,24 +311,12 @@ impl<'a> YieldScope<'a> {
         }
     }
 
-    fn clear(&mut self) {
-        self.yield_expressions.clear();
-    }
-
     fn yields_excessively(&self) -> bool {
         self.yield_expressions.len() > 1
     }
-
-    fn add_yield(&mut self, expr: &'a Expr) {
-        self.yield_expressions.push(expr);
-    }
-
-    fn set_does_return(&mut self, value: bool) {
-        self.does_return = value;
-    }
 }
 
-impl<'a> source_order::SourceOrderVisitor<'a> for YieldTracker<'a> {
+impl<'a, 'b> source_order::SourceOrderVisitor<'a> for YieldTracker<'a, 'b> {
     fn enter_node(&mut self, node: AnyNodeRef<'a>) -> source_order::TraversalSignal {
         match node {
             AnyNodeRef::StmtFor(_)
@@ -426,7 +328,7 @@ impl<'a> source_order::SourceOrderVisitor<'a> for YieldTracker<'a> {
                 // Track for primary control flow structures
                 // Optional branches like else/finally clauses are handled in leave_node
                 // Except is handled in leave node to maintain logical locality
-                self.push_scope(YieldScope::new());
+                self.push_scope();
             }
             _ => {}
         }
@@ -447,8 +349,7 @@ impl<'a> source_order::SourceOrderVisitor<'a> for YieldTracker<'a> {
                 self.handle_exclusive_branches(branch_count);
             }
             AnyNodeRef::StmtFor(_) | AnyNodeRef::StmtWhile(_) => {
-                let (else_yields, else_returns) =
-                    self.pop_scope().expect("Missing loop else scope");
+                let (else_yields, _) = self.pop_scope().expect("Missing loop else scope");
                 let (body_yields, _body_returns) =
                     self.pop_scope().expect("Missing loop body scope");
 
@@ -460,10 +361,6 @@ impl<'a> source_order::SourceOrderVisitor<'a> for YieldTracker<'a> {
                     }
                 }
                 self.propagate_yields(&else_yields);
-                if else_returns {
-                    // If else returns, don't propagate yield count
-                    self.clear_scope_yields();
-                }
             }
             _ => {}
         }
@@ -482,7 +379,7 @@ impl<'a> source_order::SourceOrderVisitor<'a> for YieldTracker<'a> {
         match stmt {
             ast::Stmt::Return(_) => {
                 if let Some(scope) = self.scopes.last_mut() {
-                    scope.set_does_return(true);
+                    scope.does_return = true;
                 }
             }
             ast::Stmt::FunctionDef(nested) => {
@@ -513,7 +410,7 @@ impl<'a> source_order::SourceOrderVisitor<'a> for YieldTracker<'a> {
                 if self.enter_node(node).is_traverse() {
                     self.visit_body(body);
                     for clause in elif_else_clauses {
-                        self.push_scope(YieldScope::new());
+                        self.push_scope();
                         self.visit_elif_else_clause(clause);
                     }
                     self.leave_node(node);
@@ -532,13 +429,13 @@ impl<'a> source_order::SourceOrderVisitor<'a> for YieldTracker<'a> {
                 if self.enter_node(node).is_traverse() {
                     self.visit_body(body);
                     for handler in handlers {
-                        self.push_scope(YieldScope::new());
+                        self.push_scope();
                         self.visit_except_handler(handler);
                     }
 
-                    self.push_scope(YieldScope::new());
+                    self.push_scope();
                     self.visit_body(orelse);
-                    self.push_scope(YieldScope::new());
+                    self.push_scope();
                     self.visit_body(finalbody);
                     self.leave_node(node);
                 }
```

</details>

My main thinking boils down to this: we're already using the `Scope` pushing and popping to manage scopes, and we're calling `add_yield` every time we encounter a `yield` expression, so it seems to me that we shouldn't have to do any post-processing on the nodes, we just push a `yield` every time we encounter it and emit a diagnostic if we've already seen too many in this `Scope`. You can see this in my `add_yield` implementation in the patch, which I think you already had mostly the same version of, I just tried to use it a bit more aggressively, I think:

```rust
    fn add_yield(&mut self, expr: &'a Expr) {
        if let Some(scope) = self.scopes.last_mut() {
            scope.yield_expressions.push(expr);
            if scope.yields_excessively() {
                self.emit_violation(expr);
            }
        }
    }
```

If we call this every time we encounter a `yield` expression, and we've set up our `Scope` stack correctly, won't we automatically emit a violation in each of the correct places?

I could be wrong and just confused about all of the various `emit_*` and `handle_*` and `propagate` functions, but intuitively this is what I was expecting.

In any case, the code I deleted is not being tested, so we may need more extensive test coverage to make sure it's actually needed.

I'm sorry to keep dragging this out, I was hoping to approve and merge this today. I'm happy to keep playing with the code until I convince myself it's correct if I'm off base here.

maxmynter committed:
(test) Yield accumulation following loop break

maxmynter commented on crates/ruff_linter/src/rules/ruff/rules/multiple_yields_contextmanager.rs:
```diff
@@ -0,0 +1,549 @@
+    }
+
+    fn clear_scope_yields(&mut self) {
+        self.scopes
+            .last_mut()
+            .expect("Missing current scope for clearing yields")
+            .clear();
```
Test added in [e0ba97b](https://github.com/astral-sh/ruff/pull/17281/commits/e0ba97b9f26f7f44573361e10a4c137e746502ef)

maxmynter committed:
(refactor) Only clear yield scopes after returning loop else

maxmynter committed:
(refactor) Reporting simplifications

maxmynter committed:
(test) try/except with return yield accumulation

maxmynter committed:
(tests) yield in finally after returning except

maxmynter committed:
(test) multiple yield in case arm

maxmynter committed:
(refactor) Don't handle returning branches in subsequent control flow

maxmynter committed:
(wip, tests) yield preceeding complex control flow

maxmynter committed:
(refactor) Simplify traversal

maxmynter committed:
(tests) fix update snapshots

maxmynter committed:
(chore) Cleanup

maxmynter committed:
(fixup) Eliminate false positives in lamba/generator

maxmynter committed:
(fixup, tests) Loop else handling

maxmynter committed:
(fixup) Replace, not accumulate

the yield counts are intitialised with the cummulative counts.
They should be assigned, not added to the outer scope yield count.

maxmynter committed:
Streamline if/else and match/case

maxmynter committed:
(chore) clean comments

maxmynter committed:
(chore) Clippy

maxmynter committed:
(chore) Align comments & test case naming

maxmynter commented:
Thank @ntBre  you for being pedantic. I'm learning a lot here. Plus, in this case it wasn't, I found edge cases this didn't handle before. And it should definitely not be possible to delete core logic without tests breaking. 

So what changed? 
- **Removed Tracker / Reporter Seperation**: Embedded `Checker` directly and removed intermediate `violations` storing
- **Simplified scope management**: Switched from tracking expression references to simple yield counts 
- **Reworked visitor logic**: Now only relies on `visit_stmt` and `visit_expr`, removed complex path-building methods and ~170 lines
- **Added 18 more test cases** including the loop-else-break scenario
- **Fixed false positives**: Now correctly ignores `yield`s in `lambda` and generators

I think the complexity around `try-except` is warranted. Mainly due to the fact that `finally` executes even when previous `try` or `except` branches `return`. We need to keep track of this state to know if we should continue to accumulate `yields` after the finally or if the branch is guarded by a `return`.

maxmynter committed:
Merge branch 'main' into ctxmg-yield-once

maxmynter committed:
Set preview
